File: mira-core/mira-ingest/db/manufacturer_normalize.py
"""Ingest-side manufacturer name normalization (issue #1596).

This module mirrors ``mira-crawler/ingest/manufacturer_normalize.py`` so the
``mira-core/mira-ingest`` service applies the SAME OCR/variant collapse at its
own DB write boundary (``db/neon.py`` ``insert_knowledge_entry`` /
``insert_knowledge_entries_batch``). The Hub KB manufacturer catalog is
computed by ``GROUP BY knowledge_entries.manufacturer``; without this, OCR /
extraction variants ("Alien-Bradley", "Cofemo", "Orldndo Rigging",
"Deshaco") each mint their own catalog row and fragment one real vendor.

The ``OCR_VARIANT_ALIASES`` map MUST stay identical to the crawler's: a
cross-service consistency test asserts the two maps match, so a vendor groups
the same way no matter which service ingested the chunk. This module is the
deterministic seed-map layer only; the crawler's fuzzy *proposer* is not
needed here (the ingest write boundary applies the seed map, nothing more).
"""

from __future__ import annotations

from dataclasses import dataclass

# Curated manufacturer-variant -> canonical map. Keys are matched
# case-insensitively with internal whitespace collapsed (see ``normalize_
# manufacturer``). MUST stay identical to the crawler's copy in
# ``mira-crawler/ingest/manufacturer_normalize.py`` (guarded by a
# cross-service consistency test). Seeded from issue #1596.
OCR_VARIANT_ALIASES: dict[str, str] = {
    "allen-bradley": "Rockwell Automation",
    "allen bradley": "Rockwell Automation",
    "alien-bradley": "Rockwell Automation",
    "alien bradley": "Rockwell Automation",
    "cofemo": "Coffing",
    "cofing": "Coffing",
    "cottins": "Coffing",
    "orldndo rigging": "Orlando Rigging",
    "deshaco": "Deshazo",
    "desha": "Deshazo",
    "deshao": "Deshazo",
    "deshazzo": "Deshazo",
}


@dataclass(frozen=True)
class NormalizedManufacturer:
    """Result of normalizing a raw manufacturer string."""

    canonical: str  # the string to store / build the catalog row from
    method: str  # "alias" | "identity"
    confidence: float  # 1.0 for deterministic alias/identity
    raw: str  # the original input


def normalize_manufacturer(raw: str | None) -> NormalizedManufacturer:
    """Collapse an OCR/extraction manufacturer variant to its canonical
    spelling, or pass an unknown vendor through unchanged.

    Whitespace is always trimmed and internally collapsed. None / blank input
    yields an empty canonical (the ``manufacturer`` column is nullable; the
    write boundary stores ``None`` for an empty canonical).
    """
    if not raw or not raw.strip():
        return NormalizedManufacturer(
            canonical="", method="identity", confidence=1.0, raw=raw or ""
        )

    key = " ".join(raw.lower().split())
    canonical = OCR_VARIANT_ALIASES.get(key)
    if canonical is not None:
        return NormalizedManufacturer(canonical=canonical, method="alias", confidence=1.0, raw=raw)

    # Unknown vendor — pass through with whitespace cleaned only.
    cleaned = " ".join(raw.split())
    return NormalizedManufacturer(canonical=cleaned, method="identity", confidence=1.0, raw=raw)

File: mira-hub/src/lib/manufacturerNormalize.ts
/**
 * Manufacturer name normalization at the Hub's KB write boundary (issue #1596).
 *
 * OCR / extraction noise fragments one real vendor into several catalog rows —
 * "Alien-Bradley" vs "Allen-Bradley", "Cofemo"/"Cofing"/"Cottins" vs "Coffing",
 * "Orldndo Rigging" vs "Orlando Rigging", "Deshaco"/"Desha"/... vs "Deshazo".
 * Because the Hub manufacturer catalog is `GROUP BY knowledge_entries.manufacturer`,
 * each variant mints its own catalog row. Collapsing variants at the insert
 * boundary keeps the catalog (and BM25 grouping) coherent.
 *
 * This is the TypeScript mirror of the crawler's
 * `mira-crawler/ingest/manufacturer_normalize.py::normalize_manufacturer`.
 * Semantics MUST stay in lockstep — the alias map lives in
 * `manufacturer-aliases.json` so a cross-surface consistency test can read it.
 */
import aliases from "./manufacturer-aliases.json";

const OCR_VARIANT_ALIASES: Record<string, string> = aliases;

/** Lowercase + collapse internal whitespace for stable lookup. Mirrors the
 * Python `_norm_key`. */
function normKey(value: string): string {
  return value.toLowerCase().split(/\s+/).filter(Boolean).join(" ");
}

export function normalizeManufacturer(
  raw: string | null | undefined,
): { canonical: string; method: "alias" | "identity" } {
  if (!raw || !raw.trim()) {
    return { canonical: "", method: "identity" };
  }

  const key = normKey(raw);
  const canonical = OCR_VARIANT_ALIASES[key];
  if (canonical !== undefined) {
    return { canonical, method: "alias" };
  }

  // Unknown vendor — pass through with whitespace cleaned only. We do NOT
  // impose a canonical of our own (matches the Python "Divergence safety").
  return { canonical: raw.trim().split(/\s+/).join(" "), method: "identity" };
}
